async-tool-docs-ag-2556 (#85)
agno-agi/agno-docs: tools/async-tools.mdx

---
title: Async Tools
---

> Agno Agents can run multiple tools concurrently, allowing you to provide the model with multiple tool call results efficiently, especially when they involve time-consuming operations. This improves responsiveness and reduces overall execution time.

Reviewer: ```suggestion
Agno Agents can execute multiple tools concurrently, allowing you to process function calls that the model makes efficiently. This is especially valuable when the functions involve time-consuming operations. It improves responsiveness and reduces overall execution time.
```

Here is an example:

```python async_tools.py
import asyncio

from agno.agent import Agent
from agno.models.openai import OpenAIChat
from agno.tools.duckduckgo import DuckDuckGoTools

agent = Agent(
    model=OpenAIChat(id="gpt-4o"),
    tools=[DuckDuckGoTools()],
    show_tool_calls=True,
    markdown=True,
    debug_mode=True,

> )

Reviewer: I think a better example would be a custom tool with async. Like we have in cookbooks.

Also add a link to the cookbooks.

asyncio.run(agent.aprint_response("Whats happening in UK and in USA?"))
```

Run the Agent:

```bash
pip install -U agno openai

export OPENAI_API_KEY=***

python async_tools.py
```

How to use:

1. Provide your Agent with a list of tools, preferably asynchronous for optimal performance. However, standard functions can also be used since they will execute concurrently on a separate thread.

> 2. Run the Agent using either the `arun` or `aprint_response` function, enabling efficient execution of tool calls.

Reviewer: ```suggestion
2. Run the Agent using either the `arun` or `aprint_response` function, enabling concurrent execution of tool calls.
```

<Note>

>   Concurrent execution of tools depends on the model requesting multiple tool
>   calls at once.

Reviewer: ```suggestion
  Concurrent execution of tools requires a model that is able to request multiple tool
  calls at once.
```

</Note>

> In this example, `gpt-4o` makes two simultaneous tool calls to `DuckDuckGoTools`. Typically, these calls would execute sequentially, causing delays. However, with Agno's `aprint_response` function, they run concurrently, significantly improving execution time.

Reviewer: ```suggestion
In this example, `gpt-4o` makes two simultaneous tool calls to `DuckDuckGoTools`. Normally these tool calls would execute sequentially, causing delays. However, using the `aprint_response` function, they run concurrently, improving execution time.
```
